server/src/main/java/org/elasticsearch/search/fetch/subphase/MatchedQueriesFetchSubPhase.java

```diff
@@ -18,82 +18,56 @@
  */
 package org.elasticsearch.search.fetch.subphase;
 
-import org.apache.lucene.index.IndexReader;
-import org.apache.lucene.index.LeafReaderContext;
-import org.apache.lucene.index.ReaderUtil;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Matches;
 import org.apache.lucene.search.Query;
 import org.apache.lucene.search.ScoreMode;
-import org.apache.lucene.search.ScorerSupplier;
 import org.apache.lucene.search.Weight;
-import org.apache.lucene.util.Bits;
-import org.elasticsearch.ExceptionsHelper;
-import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.index.query.NamedQuery;
 import org.elasticsearch.search.fetch.FetchSubPhase;
-import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.search.internal.SearchContext;
-import org.elasticsearch.search.internal.SearchContext.Lifetime;
 
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
 
 public final class MatchedQueriesFetchSubPhase implements FetchSubPhase {
 
     @Override
-    public void hitsExecute(SearchContext context, SearchHit[] hits) {
-        if (hits.length == 0 ||
-            // in case the request has only suggest, parsed query is null
-            context.parsedQuery() == null) {
+    public void hitExecute(SearchContext context, HitContext hitContext) throws IOException {
```

Comment:
I think we should implement `hitsExecute`, sorting by doc ids provide some locality and ensure that we'll read the inverted lists forward. This would also simplify the reuse of shared objects (weight).

Reply:
I'm not sure that locality will help here, given that each call to Matches pulls a new iterator for each term?

Reply:
>  Matches pulls a new iterator for each term?

This is something that can change in the future but my main point here is that we'll read the `_source/stored fields`, postings and terms vector so we should try to read forward in order to make the filesystem cache happy. This should not change the perf drastically but I am also not a fan of this API that can be misleading since only one of the function should be implemented. We wanted to remove `hitExecute` to make it less trappy but I don't remember why we've decided to keep it.